### PR TITLE
Add a compile optio --ignore-attention-mask

### PR DESCRIPTION
**Pull request by tungld**
This pull request adds a compile option `--ignore-attention-mask=arg-index` where `arg-index` is an integer number indicating the index of the attention_mask input in the model (the index is often 1 in bert-like models).

When this option is enabled, the masking computation in attention_layer are ignored.
For example, when using `--ignore-attention-mask=arg-index`, the self-attention `softmax(Q * K^T * scale + mask) * V` becomes `softmax(Q * K^T * scale) * V` where `+ mask` is removed.

### CODE REVIEW

**Comment by chentong319** [comment]
Since we cannot really safely check the condition to remove the targeted AddOp, can we have a general transformation? For example,  "--replace-op-with-its-operand=0:/encoder/layer.*/attention/self/Add"where 0 is the operand to use, the string is the regular expression for op_name.

**Comment by chentong319** [comment]
For multiple zaiu, the batch size has to be larger than one, this transformation cannot be used safely. Not sure the impact on application if this transformation is forced.

**Comment by tungld** [comment]
> Since we cannot really safely check the condition to remove the targeted AddOp, can we have a general transformation? For example, "--replace-op-with-its-operand=0:/encoder/layer.*/attention/self/Add"where 0 is the operand to use, the string is the regular expression for op_name.

Yes, using `--replace-op-with-its-operand` would indeed be more specific and flexible. My only concern is that it might be a bit advanced for typical users, e.g. they need to know node names and how this option is related to BERT-like models. If you feel strongly that this transformation is the better approach, I'm happy to make the change. Please let me know your preference, and I’ll proceed accordingly.

> For multiple zaiu, the batch size has to be larger than one, this transformation cannot be used safely. Not sure the impact on application if this transformation is forced.

You are right, I will add a sentence in the option's comment to warn about this situation.

**Comment by chentong319** [comment]
My concern of the current transformation is that the pattern does not guarantee the correctness of the transformation. How about adding a dump at compile time  when an Add is removed? Then we can use this option without worrying about unintended rewrite.
Is it true that the Op creating input of Softmax can be removed if it can be proved that op does not change the order. In this case, adding all 1 (or any same value) will not change the order of values in the tensor.

File: src/Dialect/ONNX/Transforms/IgnoreAttentionMask.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    Value mask, mmOutput;
+    if (A.getDefiningOp<ONNXMatMulOp>()) {
+      mask = B;
+      mmOutput = A;
+    } else if (B.getDefiningOp<ONNXMatMulOp>()) {
+      mask = A;
+      mmOutput = B;
+    } else
+      return failure();
+
```
Comment (chentong319) [inline src/Dialect/ONNX/Transforms/IgnoreAttentionMask.cpp]: Do we need to check whether the AddOp is MD broadcast?

Comment (tungld) [reply src/Dialect/ONNX/Transforms/IgnoreAttentionMask.cpp]: This op in the bert model is actually broadcasting.

**Comment by tungld** [comment]
> My concern of the current transformation is that the pattern does not guarantee the correctness of the transformation. How about adding a dump at compile time when an Add is removed? Then we can use this option without worrying about unintended rewrite.

I understand your concern now. Let me switch to your suggestion of `"--replace-op-with-its-operand=0:/encoder/layer.*/attention/self/Add"` to directly rewrite the op we are targeting.

> Is it true that the Op creating input of Softmax can be removed if it can be proved that op does not change the order. In this case, adding all 1 (or any same value) will not change the order of values in the tensor.

Actually, the mask tensor here includes either 0 or inf. `1` and `0` in the model input `attention_mask` becomes `0` and `inf`, respectively, after several processing.

**Comment by tungld** [comment]
@chentong319 I changed to using the option `--replace-op-with-its-operand`. Please take a look again. Thanks for your suggestion!

File: src/Dialect/ONNX/Transforms/ReplaceOpWithItsOperand.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      return WalkResult::advance();
+
+    // Match node name. Return when found the first one.
+    for (std::string IdNodenameRegex : this->nodeNameRegexList) {
+      size_t pos = IdNodenameRegex.find(":");
+      std::string inputID = IdNodenameRegex.substr(0, pos);
+      std::string nodeNameRegex = IdNodenameRegex.substr(pos + 1);
+      if (std::regex_match(opNodeName.str(), std::regex(nodeNameRegex))) {
+        int64_t id = std::stoi(inputID);
+        op->getResult(0).replaceAllUsesWith(op->getOperand(id));
```
Comment (chentong319) [inline src/Dialect/ONNX/Transforms/ReplaceOpWithItsOperand.cpp:87]: Do we want to replace all the Add in different layer with one regex like "/encoder/layer.*/attention/self/Add"?

Comment (tungld) [reply src/Dialect/ONNX/Transforms/ReplaceOpWithItsOperand.cpp:87]: Yes, the current code supports that.

**Comment by chentong319** [comment]
Why your build failed in Jenkins Linux s390 again? Is the s390 Jenkins down?

**Comment by tungld** [comment]
> Why your build failed in Jenkins Linux s390 again? Is the s390 Jenkins down?

Just merged in the fix for protobuf on s390x. Hope it works now.
